File: interchange/clients/js/interchange-client.mjs
#!/usr/bin/env node
import { createHash, generateKeyPairSync, sign, verify } from 'node:crypto';
import { readFileSync } from 'node:fs';
import { fileURLToPath, pathToFileURL } from 'node:url';
import path from 'node:path';

export function stableStringify(value) {
  if (value === null || typeof value !== 'object') return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(stableStringify).join(',')}]`;
  return `{${Object.keys(value)
    .sort()
    .map((key) => `${JSON.stringify(key)}:${stableStringify(value[key])}`)
    .join(',')}}`;
}

export function hashRef(value) {
  return `sha256:${createHash('sha256').update(stableStringify(value ?? {})).digest('hex')}`;
}

export function paramsWithoutAuth(params = {}) {
  const { auth: _auth, ...rest } = params || {};
  return rest;
}

export function canonicalPostPinMessage({
  method,
  relationshipId,
  remoteOrigin,
  nonce,
  timestamp,
  params,
}) {
  return [
    method,
    relationshipId,
    remoteOrigin,
    nonce,
    String(timestamp),
    hashRef(params || {}),
  ].join('\n');
}

export function challengeResponseHashRef({ challenge, body }) {
  return hashRef({ challenge, body: body && typeof body === 'object' ? body : {} });
}

export function createPilotAgentCard({
  name,
  agentId,
  agentUrl,
  keyId,
  publicKeyDerBase64,
}) {
  return {
    schema: 'agoragentic.agent-card.v1',
    name,
    agent_id: agentId,
    url: agentUrl,
    description: 'Partner-controlled A2A federation pilot card for Agoragentic Interchange. Protocol pilot only; not an external-demand claim.',
    extensions: {
      'agoragentic:federation': {
        schema: 'agoragentic.agent-federation.v1',
        key_id: keyId,
        public_key_der_base64: publicKeyDerBase64,
        capability_exchange: true,
        federation_consent: true,
        trust_model: 'key_control_tofu_not_identity',
      },
    },
  };
}

export function generateDemoKeyPair() {
  const { publicKey, privateKey } = generateKeyPairSync('ed25519');
  return {
    publicKey,
    privateKey,
    publicKeyDerBase64: publicKey.export({ type: 'spki', format: 'der' }).toString('base64'),
  };
}

export function signMessageBase64(privateKey, message) {
  return sign(null, Buffer.from(message, 'utf8'), privateKey).toString('base64');
}

export function verifyMessageBase64(publicKey, message, signatureBase64) {
  return verify(null, Buffer.from(message, 'utf8'), publicKey, Buffer.from(signatureBase64, 'base64'));
}

function loadVectors() {
  const here = path.dirname(fileURLToPath(import.meta.url));
  const vectorPath = path.resolve(here, '../../conformance/vectors.json');
  return JSON.parse(readFileSync(vectorPath, 'utf8'));
}

export function runSelfTest() {
  const vectors = loadVectors();
  const failures = [];

  for (const vector of vectors.post_pin_signing || []) {
    const paramsHash = hashRef(vector.params_without_auth);
    const message = canonicalPostPinMessage({
      method: vector.method,
      relationshipId: vector.relationship_id,
      remoteOrigin: vector.remote_origin,
      nonce: vector.nonce,
      timestamp: vector.timestamp,
      params: vector.params_without_auth,
    });
    const messageHex = Buffer.from(message, 'utf8').toString('hex');
    if (paramsHash !== vector.expected_params_hash) failures.push(`${vector.id}: params hash mismatch`);
    if (message !== vector.expected_canonical_message) failures.push(`${vector.id}: canonical message mismatch`);
    if (messageHex !== vector.expected_canonical_message_utf8_hex) failures.push(`${vector.id}: UTF-8 hex mismatch`);
  }

  for (const vector of vectors.challenge_response || []) {
    const challengeHash = challengeResponseHashRef({ challenge: vector.challenge, body: vector.body });
    if (challengeHash !== vector.expected_challenge_hash) failures.push(`${vector.id}: challenge hash mismatch`);
  }

  const demo = generateDemoKeyPair();
  const sample = vectors.post_pin_signing?.[0];
  if (sample) {
    const message = canonicalPostPinMessage({
      method: sample.method,
      relationshipId: sample.relationship_id,
      remoteOrigin: sample.remote_origin,
      nonce: sample.nonce,
      timestamp: sample.timestamp,
      params: sample.params_without_auth,
    });
    const signature = signMessageBase64(demo.privateKey, message);
    if (!verifyMessageBase64(demo.publicKey, message, signature)) failures.push('demo ed25519 signature failed');
  }

  return {
    ok: failures.length === 0,
    failures,
    vectors_checked: {
      post_pin_signing: vectors.post_pin_signing?.length || 0,
      challenge_response: vectors.challenge_response?.length || 0,
    },
    note: 'No network calls, payments, trust mutations, or registry submissions were made.',
  };
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  const result = runSelfTest();
  console.log(JSON.stringify(result, null, 2));
  if (!result.ok) process.exitCode = 1;
}

File: interchange/clients/python/interchange_client.py
#!/usr/bin/env python3
"""No-network Interchange v0 reference helpers.

This module intentionally implements only deterministic canonicalization and
message construction. It does not submit requests, read private keys, sign
payments, or mutate Agoragentic trust state.
"""

from __future__ import annotations

import argparse
import hashlib
import json
from pathlib import Path
from typing import Any, Dict


def stable_stringify(value: Any) -> str:
    return json.dumps(value, sort_keys=True, separators=(",", ":"), ensure_ascii=False)


def hash_ref(value: Any) -> str:
    payload = {} if value is None else value
    return "sha256:" + hashlib.sha256(stable_stringify(payload).encode("utf-8")).hexdigest()


def params_without_auth(params: Dict[str, Any]) -> Dict[str, Any]:
    return {key: value for key, value in (params or {}).items() if key != "auth"}


def params_for_post_pin_signing(*, method: str, params: Dict[str, Any]) -> Dict[str, Any]:
    without_auth = params_without_auth(params)
    if method == "federation/follow-referral":
        return {
            "relationship_id": without_auth.get("relationship_id"),
            "remote_origin": without_auth.get("remote_origin"),
            "referral_id": without_auth.get("referral_id"),
        }
    return without_auth


def canonical_post_pin_message(
    *,
    method: str,
    relationship_id: str,
    remote_origin: str,
    nonce: str,
    timestamp: Any,
    params: Dict[str, Any],
) -> str:
    return "\n".join(
        [
            method,
            relationship_id,
            remote_origin,
            nonce,
            str(timestamp),
            hash_ref(params or {}),
        ]
    )


def challenge_response_hash_ref(*, challenge: str, body: Dict[str, Any]) -> str:
    return hash_ref({"challenge": challenge, "body": body if isinstance(body, dict) else {}})


def create_pilot_agent_card(
    *,
    name: str,
    agent_id: str,
    agent_url: str,
    key_id: str,
    public_key_der_base64: str,
) -> Dict[str, Any]:
    return {
        "schema": "agoragentic.agent-card.v1",
        "name": name,
        "agent_id": agent_id,
        "url": agent_url,
        "description": "Partner-controlled A2A federation pilot card for Agoragentic Interchange. Protocol pilot only; not an external-demand claim.",
        "extensions": {
            "agoragentic:federation": {
                "schema": "agoragentic.agent-federation.v1",
                "key_id": key_id,
                "public_key_der_base64": public_key_der_base64,
                "capability_exchange": True,
                "federation_consent": True,
                "trust_model": "key_control_tofu_not_identity",
            }
        },
    }


def _load_vectors() -> Dict[str, Any]:
    vector_path = Path(__file__).resolve().parents[2] / "conformance" / "vectors.json"
    return json.loads(vector_path.read_text(encoding="utf-8"))


def run_self_test() -> Dict[str, Any]:
    vectors = _load_vectors()
    failures = []

    for vector in vectors.get("post_pin_signing", []):
        params_to_sign = vector.get("params_to_sign") or params_for_post_pin_signing(
            method=vector["method"],
            params=vector.get("wire_params") or vector["params_without_auth"],
        )
        params_hash = hash_ref(params_to_sign)
        message = canonical_post_pin_message(
            method=vector["method"],
            relationship_id=vector["relationship_id"],
            remote_origin=vector["remote_origin"],
            nonce=vector["nonce"],
            timestamp=vector["timestamp"],
            params=params_to_sign,
        )
        message_hex = message.encode("utf-8").hex()
        if params_hash != vector["expected_params_hash"]:
            failures.append(f"{vector['id']}: params hash mismatch")
        if message != vector["expected_canonical_message"]:
            failures.append(f"{vector['id']}: canonical message mismatch")
        if message_hex != vector["expected_canonical_message_utf8_hex"]:
            failures.append(f"{vector['id']}: UTF-8 hex mismatch")

    for vector in vectors.get("challenge_response", []):
        challenge_hash = challenge_response_hash_ref(challenge=vector["challenge"], body=vector["body"])
        if challenge_hash != vector["expected_challenge_hash"]:
            failures.append(f"{vector['id']}: challenge hash mismatch")

    return {
        "ok": not failures,
        "failures": failures,
        "vectors_checked": {
            "post_pin_signing": len(vectors.get("post_pin_signing", [])),
            "challenge_response": len(vectors.get("challenge_response", [])),
        },
        "note": "No network calls, payments, trust mutations, or registry submissions were made.",
    }


def main() -> int:
    parser = argparse.ArgumentParser(description="Agoragentic Interchange no-network reference client")
    parser.add_argument("--self-test", action="store_true", help="Validate bundled conformance vectors")
    args = parser.parse_args()
    if not args.self_test:
        parser.print_help()
        return 0
    result = run_self_test()
    print(json.dumps(result, indent=2, sort_keys=True))
    return 0 if result["ok"] else 1


if __name__ == "__main__":
    raise SystemExit(main())
